internal/pkg/addon/addons.go

```diff
@@ -137,55 +207,55 @@ func parseTemplate(fNames []string, workloadName string, ws workspaceReader) (*c
+// parseWorkloadParameters returns the content of user-defined additional CloudFormation Parameters
 // to pass from the parent stack to Template.
 //
 // If there are addons but no parameters file defined, then returns "" and nil for error.
 // If there are multiple parameters files, then returns "" and cannot define multiple parameter files error.
 // If the addons parameters use the reserved parameter names, then returns "" and a reserved parameter error.
-func parseParameters(fNames []string, workloadName string, ws workspaceReader) (yaml.Node, error) {
+func (p *parser) parseParameters(fNames []string) (yaml.Node, error) {
 	paramFiles := filterFiles(fNames, paramsMatcher)
 	if len(paramFiles) == 0 {
 		return yaml.Node{}, nil
 	}
 	if len(paramFiles) > 1 {
-		return yaml.Node{}, fmt.Errorf("defining %s is not allowed under %s addons/", english.WordSeries(parameterFileNames, "and"), workloadName)
+		return yaml.Node{}, fmt.Errorf("defining %s is not allowed under addons/", english.WordSeries(parameterFileNames, "and"))
 	}
 	paramFile := paramFiles[0]
-	path := ws.WorkloadAddonFilePath(workloadName, paramFile)
-	raw, err := ws.ReadFile(path)
+	path := p.filePath(paramFile)
+	raw, err := p.ws.ReadFile(path)
 	if err != nil {
-		return yaml.Node{}, fmt.Errorf("read parameter file %s under %s addons/: %w", paramFile, workloadName, err)
+		return yaml.Node{}, fmt.Errorf("read parameter file %s under path %s: %w", paramFile, path, err)
 	}
 	content := struct {
 		Parameters yaml.Node `yaml:"Parameters"`
 	}{}
 	if err := yaml.Unmarshal(raw, &content); err != nil {
-		return yaml.Node{}, fmt.Errorf("unmarshal 'Parameters' in file %s under %s addons/: %w", paramFile, workloadName, err)
+		return yaml.Node{}, fmt.Errorf("unmarshal 'Parameters' in file %s under path %s: %w", paramFile, path, err)
 	}
 	if content.Parameters.IsZero() {
-		return yaml.Node{}, fmt.Errorf("must define field 'Parameters' in file %s under %s addons/", paramFile, workloadName)
+		return yaml.Node{}, fmt.Errorf("must define field 'Parameters' in file %s under path %s", paramFile, path)
 	}
-	if err := validateReservedParameters(content.Parameters, paramFile, workloadName); err != nil {
+	if err := p.validateParameters(content.Parameters, paramFile); err != nil {
 		return yaml.Node{}, err
 	}
 
 	return content.Parameters, nil
 }
 
-func validateReservedParameters(params yaml.Node, fname, workloadName string) error {
-	content := struct {
-		App  yaml.Node `yaml:"App"`
-		Env  yaml.Node `yaml:"Env"`
-		Name yaml.Node `yaml:"Name"`
-	}{}
+func validateReservedParameters(params yaml.Node, reservedKeys []string) error {
+	content := make(map[string]yaml.Node, len(reservedKeys))
+	for _, key := range reservedKeys {
+		content[key] = yaml.Node{}
+	}
 	if err := params.Decode(&content); err != nil {
-		return fmt.Errorf("decode content of parameters file %s under %s addons/", fname, workloadName)
+		return fmt.Errorf("decode content of the parameters file: %w", err)
```

Comment:
is it really a file or just the `Parameters` section in the template?

Reply:
This is decoding the `Parameters` section of the `xx.params.yaml` file (not the template), and `Parameters` is pretty much the only section that's in `xx.params.yaml` file I think. I have updated the error message (will push later) - see if you like it better!